### PR TITLE
[PyROOT] Add numba support to PyROOT

### PR DESCRIPTION
**Pull request by sudo-panda**
These changes backport the commits that add numba support in cppyy master to PyROOT. It currently supports passing objects to C++ functions and returning builtin types from the functions inside numba JIT'ed code. It also supports accessing data members and methods from class types. The [tests added](https://github.com/sudo-panda/roottest/compare/502450f83a346aa1267fc4f48b6d749fa6038611..numba) can also be used to understand the support that is provided through this PR.

## Checklist:

- [ ] Pass all python tests in `roottest` (1 test failing - `roottest-python-memory-memory`)
- [ ] Add numba test to `roottest`

### CODE REVIEW

**Comment by etejedor** [comment]
@phsft-bot build

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by vgvassilev** [comment]
@phsft-bot build!

@Axel-Naumann, can we add Baidyanath to the team so that the bot triggers automatically?

**Comment by vgvassilev** [comment]
@phsft-bot build

**Comment by Axel-Naumann** [comment]
> can we add Baidyanath to the team so that the bot triggers automatically?

Invitation sent!

**Comment by vgvassilev** [comment]
Ping. What's the fate of this?

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by Axel-Naumann** [comment]
> Ping. What's the fate of this?

Are you pinging @sudo-panda or whom? This is still WIP according to the title; let me switch this to draft mode to make this more visible. I'd hope that the failing test can be addressed before this gets reviewed?

**Comment by sudo-panda** [comment]
@phsft-bot build

**Comment by sudo-panda** [comment]
@phsft-bot build

**Comment by vepadulano** [comment]
IIUC in this PR you are porting a major new feature of upstream cppyy into ROOT's cppyy. What is the strategy here? Are you also rebasing ROOT's cppyy on upstream, or are you cherry-picking and fixing conflicts by hand?

**Comment by sudo-panda** [comment]
I have cherry-picked the commits.

**Comment by vepadulano** [comment]
Did that work as-is? I am surprised it required no changes in ROOT's cppyy?

**Comment by sudo-panda** [comment]
It required minor changes but it works.

**Comment by sudo-panda** [comment]
I will fix that commit message. Thanks for finding that!

**Comment by sudo-panda** [comment]
@phsft-bot  build

**Comment by sudo-panda** [comment]
@phsft-bot build

**Comment by vgvassilev** [comment]
The failure seems unrelated, @sudo-panda can you mark this PR as ready to review?

**Comment by vgvassilev** [comment]
> The failure seems unrelated, @sudo-panda can you mark this PR as ready to review?

Can you remove the '[WIP]' tag?

**Comment by vepadulano** [comment]
Hi, I need to find some time to review the PR. Meanwhile, I see there are not tutorials yet and I think those would be necessary before merging. I will try now to write some pseudo-code that I *imagine* this PR would make possible to write

```python
import ROOT

ROOT.gInterpreter.Declare("""

struct A{
    int mA{};
    float mB{};
}

void callfn(int a, float b)
{
   std::cout << f(a, b) << std::endl;
}

""")


@numba.njit
def myfn():
    myobj = ROOT.A(1, 2.)
    return ROOT.callfn(myobj.mA, myobj.mB)

@numba.njit
def myfn2():
    h = ROOT.TH1F()
    return h.GetMean()
```

Let me know if this (kind-of)works, if not, so that I can better understand what is the level of support for numba-jitting ROOT classes brought by this PR.
Thanks!

**Comment by sudo-panda** [comment]
Hi @vepadulano, sorry for the very late reply. This is the part that works:
```python
import ROOT, ROOT.NumbaExt, numba

ROOT.gInterpreter.Declare('''

struct A{
    A(int A, float B) : mA(A), mB(B) {}
    int mA{};
    float mB{};
};

void callfn(int a, float b)
{
   std::cout << a << ", " << b << std::endl;
}

''')


@numba.njit
def myfn(myobj):
    return ROOT.callfn(myobj.mA, myobj.mB)

#  @numba.njit
#  def myfn2(h):
#      return h.GetMean()

obj = ROOT.A(1, 2.0)
myfn(obj)
#  myfn2(ROOT.TH1F())

```

The commented-out lines do not work.

Also, I think these tests that I have added can give a better view of what works:
https://github.com/sudo-panda/roottest/compare/502450f83a346aa1267fc4f48b6d749fa6038611..numba

**Review by vepadulano — APPROVED** [review]
Congrats @sudo-panda for this very large amount of work!

I have left some comments which are more clarifications than actual change requests.

I would have liked to see a tutorial added in this PR.

I suggest we wait for the sibling roottest PR at https://github.com/root-project/roottest/pull/975 to be ready for merging, before merging this one.

File: bindings/pyroot/pythonizations/python/ROOT/_facade.py [hunk 1]
```diff
@@ -382,6 +382,22 @@ def Numba(self):
...
+        import numba
+        if not hasattr(numba, 'version_info') or numba.version_info < (0, 54):
+            raise Exception("NumbaExt requires Numba version 0.54 or higher")
+
+        import cppyy.numba_ext
+
+        # Register the NumbaExt as a namespace and return anything
+        # Both these steps are required to make ROOT find the module
+        cppdef('namespace NumbaExt {}')
+        return self._
```
[the remark this answers is not in the document]
Comment (sudo-panda) [reply bindings/pyroot/pythonizations/python/ROOT/_facade.py:402]: For the namespace one, I am not sure but the `Numba` property also has to do this so I tried it and it worked.

For the return one, it makes sense as the function is marked `@property` so it should return something.

File: bindings/pyroot/cppyy/cppyy/python/cppyy/numba_ext.py [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+
+    return cnt
+
+
+#
+# C++ instance -> Numba
+#
+@nb_ext.typeof_impl.register(cpp_types.Instance)
+def typeof_instance(val, c):
+    global scope_numbatypes
```
[the remark this answers is not in the document]
Comment (sudo-panda) [reply bindings/pyroot/cppyy/cppyy/python/cppyy/numba_ext.py:461]: See https://github.com/root-project/root/pull/11096#discussion_r1235310882

File: bindings/pyroot/cppyy/cppyy/python/cppyy/numba_ext.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+        model = nb_dm.default_manager.lookup(typ)
+        pobj = c.builder.bitcast(vptr, model.get_argument_type())
+
+        return nb_ext.NativeValue(pobj, is_error=None, cleanup=None)
+
+  # C++ object to Python proxy wrapping for returns from Numba trace
+    @nb_ext.box(ImplClassType)
+    def box_instance(typ, val, c):
+        assert not "requires object model and passing of intact object, not memberwise copy"
+        global cppyy_from_voidptr
```
[the remark this answers is not in the document]
Comment (sudo-panda) [reply bindings/pyroot/cppyy/cppyy/python/cppyy/numba_ext.py:438]: See https://github.com/root-project/root/pull/11096#discussion_r1235310882

File: bindings/pyroot/cppyy/cppyy/python/cppyy/numba_ext.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+                members = [(dmi.f_name, dmi.f_nbtype) for dmi in data_members]
+                nb_dm.models.StructModel.__init__(self, dmm, fe_type, members)
+
+    else:
+        assert not "unknown qualified type"
+
+  # Python proxy unwrapping for arguments into the Numba trace
+    @nb_ext.unbox(ImplClassType)
+    def unbox_instance(typ, obj, c):
+        global cppyy_as_voidptr
```
[the remark this answers is not in the document]
Comment (sudo-panda) [reply bindings/pyroot/cppyy/cppyy/python/cppyy/numba_ext.py:422]: See https://github.com/root-project/root/pull/11096#discussion_r1235310882

File: bindings/pyroot/cppyy/cppyy/python/cppyy/numba_ext.py [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+        assert not "unknown qualified type"
+
+    return context.cppyy_currentcall_this
+
+
+scope_numbatypes = (dict(), dict())
+
+@nb_ext.typeof_impl.register(cpp_types.Scope)
+def typeof_scope(val, c, q = Qualified.default):
+    global scope_numbatypes
```
[the remark this answers is not in the document]
Comment (vepadulano) [reply bindings/pyroot/cppyy/cppyy/python/cppyy/numba_ext.py:321]: What happens if you remove the `global` keywords there? The caches are at module level, so you should be able to access and modify them inside the functions. Can you try?

[the remark this answers is not in the document]
Comment (sudo-panda) [reply bindings/pyroot/cppyy/cppyy/python/cppyy/numba_ext.py:321]: I believe I have found a cleaner way to do this. Can you check if [this](https://github.com/sudo-panda/root/blob/00d67a43bee72688ef9cca69c3c8f1ee3d266222/bindings/pyroot/pythonizations/python/ROOT/_facade.py#L399-L400) is good?

[the remark this answers is not in the document]
Comment (vepadulano) [reply bindings/pyroot/cppyy/cppyy/python/cppyy/numba_ext.py:321]: `global` keyword should *never* be used.

The SO answer you refer to is correct but only partially:

```python
cache = {}

def foo():
    cache["foo"] = 1

def bar():
    cache["bar"] = 2

print(f"{cache=}")
foo()
bar()
print(f"{cache=}")
```
```
$: python repro.py
cache={}
cache={'foo': 1, 'bar': 2}
```

**Comment by vgvassilev** [comment]
One thing to note here is that this code live as is in cppyy upstream. Changing it would cause more source of divergence and would make it very hard to backport features. I am almost certain non-critical changes won’t make it back upstream.

My general recommendation is that unless there is something critical to change we should keep things as they are. Otherwise we will suffer in the near future.

**Comment by Axel-Naumann** [comment]
> I am almost certain that non-critical changes won’t make it back upstream.

I think all of us appreciate the time and brain power Vincenzo invests in improving the code. I bet it's useful to propose changes coming out of this upstream where relevant, @sudo-panda ?

**Comment by sudo-panda** [comment]
> > I am almost certain that non-critical changes won’t make it back upstream.
> 
> I think all of us appreciate the time and brain power Vincenzo invests in improving the code. I bet it's useful to propose changes coming out of this upstream where relevant, @sudo-panda ?

Hi @Axel-Naumann ,

I am really grateful to @vepadulano for investing his precious time in reviewing this PR and suggesting changes to improve it. If not for his suggestions about the `namespace NumbaExt` code, I wouldn't have been able to simplify it.

On the other hand, @vgvassilev  is right about the global statements being a non-critical change and wouldn't be accepted upstream. And I am sure when he said that, he didn't mean that Vincenzo's efforts were wasteful.

Finally, AFAIU Vincenzo thinks that global statements are unnecessary because the code acts in the same way whether you keep them or not. I have a few counterpoints to this:

1. It gives programmers an idea about where the variable was initialized before use. Modifying the code provided by Vincenzo:

    ```python
    cache = {}

    def foo():
        .....
        .... Many lines of code ...
        .....
        cache["foo"] = 1

    .....
    .... Many lines of code ...
    .....

    def bar():
        .....
        .... Many lines of code ...
        .....
        cache["bar"] = 2
    ```

    I hope you can see it becomes difficult to understand where did cache come from. A `global cache` statement at the top of the function or before its use clarifies the intent.

2. Removing the `global` statement doesn't affect the scope of the variable. It is purely meant to [guide Python for the code block it is used in.](https://docs.python.org/3/reference/simple_stmts.html#the-global-statement)

3. ROOT already uses `global` keyword in [JupyROOT](https://github.com/root-project/root/blame/a90d76b7798dc4332425d29794089fceaebccb60/bindings/jupyroot/python/JupyROOT/helpers/utils.py#L126 ). So unless "we should never use `global`" is a new convention being followed in ROOT, I believe it doesn't seem to track.

4. As seen in this [SO answer](https://stackoverflow.com/a/4693392), Python used to differentiate between having a global statement or not. And even though the answer is more than ten years old, I experienced this firsthand when I started using Python (about four years back). So even though Python changed how it treats the variable binding in the newer versions, I would say it doesn't harm us by keeping them and might benefit us in the future.

If these counterpoints do not convince you, I will remove the global statements in the PR. :)

Sorry for the long comment.

**TL;DR:** I am thankful to Vincenzo for taking the time out of his busy schedule to review this PR. I have some counterpoints to his arguments against `global` and if he doesn't agree to them I will remove them ASAP.

**Comment by vepadulano** [comment]
Dear all,
I approved this PR a few days ago now, the 2 remaining comments were one curiosity and one suggestion regarding general best practices in programming, not strictly related to this particular piece of code. Nothing prevents this PR from merging and I believe the discussion is going further than needed at this point.
